File: Python/algorithmic_toolbox/26_longest_common_subsequence.py
import sys
from collections import namedtuple

Sequence = namedtuple("Sequence", ["seq", "length"])


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_input_data(data):
    sequences = []
    k = 0
    while k < 3:
        ix = 2 * k
        length = int(data[ix])
        sequence = tuple(convert_to_intlist(data[ix+1]))
        sequences.append(Sequence(seq=sequence, length=length))
        k += 1
    return sequences


def calc_longest_common_subsequence(data):
    sequences, lengths = zip(*data)
    s1, s2, s3 = sequences
    n1, n2, n3 = lengths

    longest_path = [[[0] * (n3+1) for _ in range(n2+1)] for _ in range(n1+1)]
    for i in range(n1):
        for j in range(n2):
            for k in range(n3):
                if s1[i] == s2[j] and s1[i] == s3[k]:
                    longest_path[i+1][j+1][k+1] = longest_path[i][j][k] + 1
                else:
                    longest_path[i+1][j+1][k+1] = max(longest_path[i][j+1][k+1],
                                                      longest_path[i+1][j][k+1],
                                                      longest_path[i+1][j+1][k])
    return longest_path[-1][-1][-1]


def main():
    data = sys.stdin.read().splitlines()
    sequences = read_input_data(data)
    result = calc_longest_common_subsequence(sequences)
    print(result)


if __name__ == "__main__":
    main()

File: Julia/algorithmic_toolbox/26-longest-common-subsequence.jl
function readinputdata()
    sequences = NamedTuple{(:seq, :length), Tuple{Vector{Int}, Int}}[]
    k = 0
    for k = 1:3
        ix = 2k
        size = parse(Int, readline())
        sequence = map(c -> parse(Int, c), split(readline()))
        push!(sequences, (seq=sequence, length=size))
    end
    sequences
end


function longestcommonsubsequence(data)
    s₁, s₂, s₃ = map(s -> s.seq, data)
    n₁, n₂, n₃ = map(s -> s.length, data)

    longestpath = zeros(Int, (n₁ + 1, n₂ + 1, n₃ + 1))
    for i in 1:n₁, j in 1:n₂, k in 1:n₃
        longestpath[i+1, j+1, k+1] = begin
            if s₁[i] == s₂[j] == s₃[k]
                longestpath[i, j, k] + 1
            else
                max(longestpath[i, j+1, k+1], longestpath[i+1, j, k+1], longestpath[i+1, j+1, k])
            end
        end
    end
    longestpath[n₁+1, n₂+1, n₃+1]
end


function main()
    data = readinputdata()
    result = longestcommonsubsequence(data)
    println(result)
end


main()
